### PR TITLE
Replace BigNumber with bigint

### LINKED ISSUE #1554: Remove BigNumber from project
In an effort to replace Ethers with Viem, we need to stop using BigNumber types throughout the app.

### PR DESCRIPTION
**Pull request by adamgall**
Closes #1554 

This is to prepare for a codebase without `ethers.js` (v5, at least).

Next will be changes to the `fractal-contracts` repo / npm package, to publish plain `abi` objects instead of ethers-flavored TypeChain objects and types.

After that will be using `viem` contract objects and plain `readContract` calls in `fractal-interface` instead of the ethers-flavored TypeChain objects and types.

After that it'll be trivial to stop using the `ethers` `provider` and `signer`, and just use `viem`'s `publicClient` and `walletClient`.

After that we'll see the full power of `viem`'s RPC batching in action, giving (I hope to bob) extreme improvements in our network traffic performance.

### CODE REVIEW

**Comment by adamgall** [comment]
# 🚨

This code very very likely breaks proposal templates, in which the JSON data on IPFS has keys named things like `bigNumberValue`. Since there are literally no more references to the string `BigNumber` in this code, then reading existing IPFS proposal templates is very very likely to fail. Need to go undo that IPFS proposal template key naming stuff specifically, and think through this a bit further.

**Review by DarksightKellar — APPROVED** [review]
Oh yeah this is looking sweet. I'm gonna go ahead and slap on that approve, I reckon any breaks, if relevant, should be addressed in a follow up

File: src/components/Proposals/ProposalSummary.tsx [hunk 1]
```diff
@@ -91,22 +90,23 @@ export default function ProposalSummary({
 
   const strategyQuorum =
     votesToken && isERC20
-      ? votingStrategy.quorumPercentage!.value.toNumber()
+      ? Number(votingStrategy.quorumPercentage!.value.toString())
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/Proposals/ProposalSummary.tsx:93]: Nah, u rite

File: src/components/pages/DaoDashboard/Activities/ActivityFreeze.tsx [hunk 2]
```diff
@@ -33,10 +33,10 @@ export function ActivityFreeze() {
   } = useFractal();
   const { t } = useTranslation('dashboard');
   const freezeProposalDeadlineDate = new Date(
-    freezeProposalCreatedTime!.add(freezeProposalPeriod!).mul(1000).toNumber(),
+    Number((((freezeProposalCreatedTime || 0n) + (freezeProposalPeriod || 0n)) * 1000n).toString()),
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/pages/DaoDashboard/Activities/ActivityFreeze.tsx:36]: I knooo but I think it's necessary here

[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/pages/DaoDashboard/Activities/ActivityFreeze.tsx:36]: Ok, got rid of one of them by removing that `.toString()`

File: src/components/Proposals/ProposalSummary.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/Proposals/ProposalSummary.tsx]: Fixed and pushed

File: src/components/DaoCreator/formComponents/AzoriusTokenDetails.tsx [hunk 3]
```diff
@@ -49,18 +49,21 @@ export function AzoriusTokenDetails(props: ICreationStepProps) {
...
     const importError = errors?.erc20Token?.tokenImportAddress;
-    if (importAddress && !importError && utils.isAddress(importAddress)) {
+    if (importAddress && !importError && isAddress(importAddress)) {
       const isVotesToken = await checkVotesToken(importAddress);
       const tokenContract = new ethers.Contract(importAddress, erc20Abi, provider);
       const name: string = await tokenContract.name();
       const symbol: string = await tokenContract.symbol();
       const decimals: number = await tokenContract.decimals();
+
+      // @todo yo what is going on here
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/DaoCreator/formComponents/AzoriusTokenDetails.tsx]: ah, I get what you're saying. thanks. I've changed my `@todo` into a `@dev` explaining this.

File: src/components/ui/links/EtherscanLinkERC20.tsx [hunk 4]
```diff
@@ -17,9 +17,7 @@ function EtherscanLinkERC20({ address, children, ...rest }: Props) {
 
   return (
     <EtherscanLinkBase
-      href={`${etherscanBaseURL}/${
-        address === ethers.constants.AddressZero ? '' : 'token/'
-      }${address}`}
+      href={`${etherscanBaseURL}/${address === zeroAddress ? '' : 'token/'}${address}`}
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/ui/links/EtherscanLinkERC20.tsx]: yeah, good call. fixed and pushed

File: src/hooks/DAO/loaders/governance/useAzoriusListeners.ts [hunk 5]
```diff
@@ -49,22 +48,24 @@ const proposalCreatedEventListener = (
       return;
     }
 
+    const typedTransactions = transactions.map(t => ({ ...t, value: t.value.toBigInt() }));
+
```
Comment (Da-Colon) [inline src/hooks/DAO/loaders/governance/useAzoriusListeners.ts:52]: Its probably something in decodeTransaction function, of which I have yet to look at. But What is the reason for this change?

File: src/hooks/DAO/loaders/governance/useAzoriusProposals.ts [hunk 6]
```diff
@@ -143,15 +143,18 @@ export const useAzoriusProposals = () => {
           const metadataEvent: ProposalMetadata = JSON.parse(proposalCreatedEvent.args.metadata);
           const decodedTransactions = await decodeTransactions(
             _decode,
-            proposalCreatedEvent.args.transactions,
+            proposalCreatedEvent.args.transactions.map(t => ({ ...t, value: t.value.toBigInt() })),
```
Comment (Da-Colon) [inline src/hooks/DAO/loaders/governance/useAzoriusProposals.ts:146]: o huh, being done here too

**Review by Da-Colon — COMMENTED** [review]
oof what a code review. Nothing beyond are earlier discussion which we have already fixed. A question about a few spots but just need to go click around the preview for approval

File: netlify/functions/tokenPrices.mts [hunk 7]
```diff
@@ -1,6 +1,6 @@
-// eslint-disable-next-line import/named
-import { Store, getStore } from '@netlify/blobs';
-import { ethers } from 'ethers';
+import { getStore } from '@netlify/blobs';
+import type { Store } from '@netlify/blobs';
```
[the remark this answers is not in the document]
Comment (adamgall) [reply netlify/functions/tokenPrices.mts:2]: Yeah, every once in a while Typescript or ESLint (not sure who) will yell at me about it. It did while I was in this file, hence the update.

File: src/hooks/DAO/loaders/governance/useAzoriusListeners.ts (hunk 5 above)
Comment (adamgall) [reply src/hooks/DAO/loaders/governance/useAzoriusListeners.ts:52]: The `transactions` object, which comes directly from the TypeChain type from the Event type from our NPM export, has a `value` property of type `BigNumber`. The code here is to massage that into a `bigint` type so it fits in with this codebase.

This alludes to the next step in this bigger process, which is removing TypeChain exports from the NPM package (or at least, not using them in this `interface` repo any more). They all use `BigNumbers` internally and give responses in `BigNumber`.

File: src/hooks/DAO/loaders/governance/useAzoriusProposals.ts (hunk 6 above)
Comment (adamgall) [reply src/hooks/DAO/loaders/governance/useAzoriusProposals.ts:146]: Yep

File: src/hooks/DAO/loaders/governance/useAzoriusListeners.ts (hunk 5 above)
Comment (Da-Colon) [reply src/hooks/DAO/loaders/governance/useAzoriusListeners.ts:52]: Ah make sense

**Review by Da-Colon — CHANGES_REQUESTED** [review]
everything is looking good. moving around the app and different DAOs I don't seen anything out of place.

In Live Testing, I'm unable to create a new DAO, or a Sub DAO. The submit button is highlighted but clicking it does nothing with no errors in console or on screen.

Not really sure where this problem might be.

File: src/components/DaoCreator/formComponents/AzoriusTokenAllocations.tsx [hunk 8]
```diff
@@ -72,7 +72,7 @@ export function AzoriusTokenAllocations(props: ICreationStepProps) {
                 const amountErrorMessage =
                   values.erc20Token.tokenSupply.value &&
                   tokenAllocationError?.amount?.value &&
-                  !tokenAllocation.amount.bigNumberValue?.isZero()
+                  (tokenAllocation.amount.bigintValue ?? 0n !== 0n)
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/DaoCreator/formComponents/AzoriusTokenAllocations.tsx:75]: You're absolutely right.

File: src/components/ui/links/EtherscanLinkERC20.tsx [hunk 9]
```diff
@@ -11,15 +11,13 @@ interface Props extends LinkProps {
 function EtherscanLinkERC20({ address, children, ...rest }: Props) {
   const { etherscanBaseURL } = useNetworkConfig();
 
-  if (!address) {
+  if (!address || address === zeroAddress) {
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/components/ui/links/EtherscanLinkERC20.tsx:14]: Why are we giving a special case to the zero address in anyway? 

Maybe, if the caller of this component wants to pass in the zero address, that's their decision, and we should just render the link as etherscan.io/token/0x000..000 and call it a day.

File: src/utils/shutter.ts [hunk 10]
```diff
@@ -17,7 +17,7 @@ export default async function encryptWithShutter(
   const is32ByteString = id.substring(0, 2) === '0x';
   const proposalId = arrayify(is32ByteString ? id : formatBytes32String(id));
 
-  const sigma = arrayify(BigNumber.from(randomBytes(32)));
```
[the remark this answers is not in the document]
Comment (adamgall) [reply src/utils/shutter.ts:20]: It indeed seems like we don't need those extra functions and castings here:

```ts
import { init, encrypt } from "@shutter-network/shutter-crypto";
import { BigNumber, utils } from "ethers";

async function encryptWithShutter(
  choice: string,
  id: string,
  randBytes: Uint8Array
): Promise<string | null> {
  await init();

  const { arrayify, hexlify, toUtf8Bytes, formatBytes32String } = utils;

  const bytesChoice = toUtf8Bytes(choice);
  const message = arrayify(bytesChoice);
  const eonPublicKey = arrayify(
    "0x0e6493bbb4ee8b19aa9b70367685049ff01dc9382c46aed83f8bc07d2a5ba3e6030bd83b942c1fd3dff5b79bef3b40bf6b666e51e7f0be14ed62daaffad47435265f5c9403b1a801921981f7d8659a9bd91fe92fb1cf9afdb16178a532adfaf51a237103874bb03afafe9cab2118dae1be5f08a0a28bf488c1581e9db4bc23ca"
  );

  const is32ByteString = id.substring(0, 2) === "0x";
  const proposalId = arrayify(is32ByteString ? id : formatBytes32String(id));

  const sigma = randBytes;

  const encryptedMessage = await encrypt(
    message,
    eonPublicKey,
    proposalId,
    sigma
  );

  return hexlify(encryptedMessage) ?? null;
}

async function encryptWithShutterBN(
  choice: string,
  id: string,
  randBytes: Uint8Array
): Promise<string | null> {
  await init();

  const { arrayify, hexlify, toUtf8Bytes, formatBytes32String } = utils;

  const bytesChoice = toUtf8Bytes(choice);
  const message = arrayify(bytesChoice);
  const eonPublicKey = arrayify(
    "0x0e6493bbb4ee8b19aa9b70367685049ff01dc9382c46aed83f8bc07d2a5ba3e6030bd83b942c1fd3dff5b79bef3b40bf6b666e51e7f0be14ed62daaffad47435265f5c9403b1a801921981f7d8659a9bd91fe92fb1cf9afdb16178a532adfaf51a237103874bb03afafe9cab2118dae1be5f08a0a28bf488c1581e9db4bc23ca"
  );

  const is32ByteString = id.substring(0, 2) === "0x";
  const proposalId = arrayify(is32ByteString ? id : formatBytes32String(id));

  const randomBytesBigNumber = BigNumber.from(randBytes);
  const sigma = arrayify(randomBytesBigNumber);

  const encryptedMessage = await encrypt(
    message,
    eonPublicKey,
    proposalId,
    sigma
  );

  return hexlify(encryptedMessage) ?? null;
}

(async () => {
  const choice = "yes";
  const id = "0xabc123";
  const randBytes = utils.randomBytes(32);

  const originalCode = await encryptWithShutterBN(choice, id, randBytes);
  const newCode = await encryptWithShutter(choice, id, randBytes);

  console.log({ originalCode });
  console.log({ newCode });
  console.log(originalCode === newCode);
})();
```

outputs:

```
➜  testing123 npx ts-node src/arrayify.ts
{
  originalCode: '0x020c32d761d4872541d02e54c75a5816c85c99338a1554d585a2b96d6da527c30101f50c68bd8ecf3b74915632cbe19b9006ca2ff68790d8fd57dae2d0aa257e0a70de0a1965e02378f9dd8b9340d32a402f696078da6d23654f91d5f87050dd0a7c06d315b7daac41f554522db11d9f18a070b0862803034ca8b60f3e8e0bb9cdd9e4e7ba47a68f004915f660cb8afbfe8d1739713ebef0e4849113d1023ea591417d5eb9e9c7430f5fad174ef3cdb195f1a738cb1feb81344478067c323256'
}
{
  newCode: '0x020c32d761d4872541d02e54c75a5816c85c99338a1554d585a2b96d6da527c30101f50c68bd8ecf3b74915632cbe19b9006ca2ff68790d8fd57dae2d0aa257e0a70de0a1965e02378f9dd8b9340d32a402f696078da6d23654f91d5f87050dd0a7c06d315b7daac41f554522db11d9f18a070b0862803034ca8b60f3e8e0bb9cdd9e4e7ba47a68f004915f660cb8afbfe8d1739713ebef0e4849113d1023ea591417d5eb9e9c7430f5fad174ef3cdb195f1a738cb1feb81344478067c323256'
}
true
```

**Comment by adamgall** [comment]
> everything is looking good. moving around the app and different DAOs I don't seen anything out of place.
> 
> In Live Testing, I'm unable to create a new DAO, or a Sub DAO. The submit button is highlighted but clicking it does nothing with no errors in console or on screen.
> 
> Not really sure where this problem might be.

@Da-Colon works for me. Can you try again?

**Review by Da-Colon — APPROVED** [review]
Weird nonce, thing that was mention in Slack, But the DAO creation has been fixed!

Was about to create parent and sub DAOs with all the numbers looking correct.

**Pull request merged**